test/common/stats/thread_local_store_test.cc

```diff
@@ -1186,5 +1195,196 @@ TEST_F(HistogramTest, ParentHistogramBucketSummary) {
+          tls_->registerThread(*dispatcher, false);
+        }
+        store_->initializeThreading(*main_dispatcher_, *tls_);
+      }));
+    }
+  }
+
+  ~ClusterShutdownCleanupStarvationTest() {
+    {
+      BlockingScope blocking_scope(1);
+      main_dispatcher_->post(blocking_scope.run([this]() {
+        store_->shutdownThreading();
+        tls_->shutdownGlobalThreading();
+        tls_->shutdownThread();
+      }));
+    }
+
+    for (Event::DispatcherPtr& dispatcher : thread_dispatchers_) {
+      dispatcher->post([&dispatcher]() { dispatcher->exit(); });
+    }
+
+    for (Thread::ThreadPtr& thread : threads_) {
+      thread->join();
+    }
+
+    main_dispatcher_->post([this]() {
+      store_.reset();
+      tls_.reset();
+      main_dispatcher_->exit();
+    });
+    main_thread_->join();
+  }
+
+  void incCounters() {
+    for (uint32_t i = 0; i < NumScopes; ++i) {
+      ScopePtr scope = store_->createScope("scope.");
+      Counter& counter = scope->counterFromStatName(my_counter_name_);
+      counter.inc();
+      uint32_t use_count = counter.use_count();
+      if ((use_count % 10000) == 0) {
+        // Run tests with "-l warning" to see this message.
+        ENVOY_LOG_MISC(warn, "{}: very high use-count: {}", counter.name(), use_count);
+      }
+    }
+  }
+
+  void workerThreadFn(uint32_t thread_index, BlockingScope& blocking_scope) {
+    thread_dispatchers_[thread_index] = api_->allocateDispatcher();
+    blocking_scope.decrementCount();
+    thread_dispatchers_[thread_index]->run(Event::Dispatcher::RunType::RunUntilExit);
+  }
+
+  void mainThreadFn(BlockingScope& blocking_scope) {
+    main_dispatcher_ = api_->allocateDispatcher();
+    blocking_scope.decrementCount();
+    main_dispatcher_->run(Event::Dispatcher::RunType::RunUntilExit);
+  }
+
+  void incCountersAllThreads() {
+    BlockingScope blocking_scope(NumThreads);
+    for (Event::DispatcherPtr& thread_dispatcher : thread_dispatchers_) {
+      thread_dispatcher->post(blocking_scope.run([this]() { incCounters(); }));
+    }
+  }
+
+  std::chrono::seconds elapsedTime() {
+    return std::chrono::duration_cast<std::chrono::seconds>(time_system_.monotonicTime() -
+                                                            start_time_);
+  }
+
+  Event::TestRealTimeSystem time_system_;
+  MonotonicTime start_time_;
+  Api::ApiPtr api_;
+  Event::DispatcherPtr main_dispatcher_;
+  std::vector<Event::DispatcherPtr> thread_dispatchers_;
+  Thread::ThreadFactory& thread_factory_;
+  std::unique_ptr<ThreadLocal::InstanceImpl> tls_;
+  Thread::ThreadPtr main_thread_;
+  std::vector<Thread::ThreadPtr> threads_;
+  StatNamePool pool_;
+  StatName my_counter_name_;
+  StatName my_counter_scoped_name_;
+};
+
+// Tests the scenario where a cluster and stat are allocated in multiple
+// concurrent threads, but after each round of allocation/free we post() an
+// empty callback to main to ensure that cross-scope thread cleanups complete.
```

Comment:
The post() itself doesn't do anything, it just forces the dispatcher to run and process callbacks? I'm a little confused why this is needed. Is the dispatcher not actually running?

Reply:
It ensures that I was able to queue up a callback and have it run, but I think I can make this more clear. Will rewrite